File: jscomp/others/bs_Map.ml
(***********************************************************************)
(*                                                                     *)
(*                                OCaml                                *)
(*                                                                     *)
(*            Xavier Leroy, projet Cristal, INRIA Rocquencourt         *)
(*                                                                     *)
(*  Copyright 1996 Institut National de Recherche en Informatique et   *)
(*  en Automatique.  All rights reserved.  This file is distributed    *)
(*  under the terms of the GNU Library General Public License, with    *)
(*  the special exception on linking described in file ../LICENSE.     *)
(*                                                                     *)
(***********************************************************************)
(** Adapted by authors of BuckleScript without using functors          *)

module N = Bs_MapDict
module A = Bs_Array

type ('key, 'id ) dict = ('key, 'id) Bs_Dict.comparable
type ('key, 'id ) cmp = ('key, 'id) Bs_Dict.cmp
module S = struct 
  type ('k,'v,'id) t = {
    cmp: ('k,'id) cmp;
    data: ('k,'v, 'id) N.t
  }
  [@@bs.deriving abstract]
end

type ('k, 'v, 'id ) t = ('k, 'v, 'id) S.t 

let ofArray (type k) (type id) data ~(dict : (k,id) dict)  =
  let module M = (val dict) in
  let cmp = M.cmp in 
  S.t ~cmp ~data:(N.ofArray ~cmp data) 


let remove m x  =   
  let cmp, odata = S.cmp m, S.data m in 
  let newData = N.remove odata x ~cmp  in
  if newData == odata then m
  else S.t ~cmp ~data:newData 

let removeMany m x =     
  let cmp, odata = S.cmp m,  S.data m in
  let newData = N.removeMany odata x ~cmp in
  S.t ~cmp  ~data:newData

let set m key d  = 
  let cmp = S.cmp m in 
  S.t ~cmp  ~data:(N.set ~cmp (S.data m) key d) 

let mergeMany m e = 
  let cmp = S.cmp m in 
  S.t ~cmp  ~data:(N.mergeMany ~cmp (S.data m) e)

let update m key f  = 
  let cmp = S.cmp m in 
  S.t ~cmp ~data:(N.update ~cmp (S.data m) key f )

let split m x =   
  let cmp = S.cmp m in 
  let (l,r),b = N.split ~cmp (S.data m) x in 
  (S.t ~cmp ~data:l, S.t ~cmp  ~data:r), b  

let merge s1 s2 f = 
  let cmp = S.cmp s1 in 
  S.t ~cmp ~data:(N.merge ~cmp  (S.data s1) (S.data s2) f)

let make (type elt) (type id) ~(dict: (elt, id) dict) =
  let module M = (val dict) in 
  S.t  ~cmp:M.cmp  ~data:N.empty

let isEmpty map = 
  N.isEmpty (S.data map)


let forEach m f = N.forEach (S.data m) f

let reduce m acc f = N.reduce (S.data m) acc f  

let every m f = N.every (S.data m) f  

let some m f = N.some (S.data m) f

let keep m f =
  S.t ~cmp:(S.cmp m) ~data:(N.keep (S.data m) f)

let partition m p =   
  let cmp = S.cmp m in 
  let l,r = N.partition (S.data m) p in 
  S.t ~cmp ~data:l, S.t ~cmp ~data:r 

let map m f = 
  S.t ~cmp:(S.cmp m) ~data:(N.map (S.data m) f)
    
let mapWithKey m  f = 
  S.t ~cmp:(S.cmp m) ~data:(N.mapWithKey (S.data m) f)

let size map = N.size (S.data map)   
let toList map = N.toList (S.data map) 
let toArray m = N.toArray (S.data m)
let keysToArray m = N.keysToArray (S.data m)
let valuesToArray m = N.valuesToArray (S.data m)
let minKey m = N.minKey (S.data m)
let minKeyUndefined m = N.minKeyUndefined (S.data m)
let maxKey m = N.maxKey (S.data m)
let maxKeyUndefined m = N.maxKeyUndefined (S.data m)    
let minimum m = N.minimum (S.data m)
let minUndefined m = N.minUndefined (S.data m) 
let maximum m = N.maximum (S.data m)
let maxUndefined m = N.maxUndefined (S.data m)

let get map x  =
  N.get ~cmp:(S.cmp map)  (S.data map) x 

let getUndefined map x = 
  N.getUndefined ~cmp:(S.cmp map) (S.data map) x

let getWithDefault map x def = 
  N.getWithDefault ~cmp:(S.cmp map) (S.data map) x  def

let getExn map x = 
  N.getExn ~cmp:(S.cmp map) (S.data map) x 

let has map x = 
  N.has ~cmp:(S.cmp map) (S.data map) x

let checkInvariantInternal m  =
  N.checkInvariantInternal (S.data m)

let eq m1 m2 veq = 
  N.eq ~kcmp:(S.cmp m1) ~veq (S.data m1) (S.data m2)

let cmp m1 m2 vcmp =
  N.cmp ~kcmp:(S.cmp m1)  ~vcmp (S.data m1) (S.data m2)

let getData = S.data
  
let getDict (type elt) (type id) (m : (elt,_,id) t) : (elt, id) dict =
  let module T = struct
    type nonrec id = id
    type nonrec t = elt
    let cmp =  S.cmp m
  end in
  (module T )
  
let packDictData (type elt) (type id) ~(dict : (elt, id) dict) ~data  =
  let module M = (val dict) in 
  S.t ~cmp:M.cmp ~data

module Int = Bs_MapInt
module String = Bs_MapString

File: lib/js/bs_Map.js
'use strict';

var Curry = require("./curry.js");
var Bs_MapDict = require("./bs_MapDict.js");

function ofArray(data, id) {
  var cmp = id[/* cmp */0];
  return {
          cmp: cmp,
          data: Bs_MapDict.ofArray(data, cmp)
        };
}

function remove(m, x) {
  var cmp = m.cmp;
  var odata = m.data;
  var newData = Bs_MapDict.remove(odata, x, cmp);
  if (newData === odata) {
    return m;
  } else {
    return {
            cmp: cmp,
            data: newData
          };
  }
}

function removeMany(m, x) {
  var cmp = m.cmp;
  var odata = m.data;
  var newData = Bs_MapDict.removeMany(odata, x, cmp);
  return {
          cmp: cmp,
          data: newData
        };
}

function set(m, key, d) {
  var cmp = m.cmp;
  return {
          cmp: cmp,
          data: Bs_MapDict.set(m.data, key, d, cmp)
        };
}

function mergeMany(m, e) {
  var cmp = m.cmp;
  return {
          cmp: cmp,
          data: Bs_MapDict.mergeMany(m.data, e, cmp)
        };
}

function updateU(m, key, f) {
  var cmp = m.cmp;
  return {
          cmp: cmp,
          data: Bs_MapDict.updateU(m.data, key, f, cmp)
        };
}

function update(m, key, f) {
  return updateU(m, key, Curry.__1(f));
}

function split(m, x) {
  var cmp = m.cmp;
  var match = Bs_MapDict.split(m.data, x, cmp);
  var match$1 = match[0];
  return /* tuple */[
          /* tuple */[
            {
              cmp: cmp,
              data: match$1[0]
            },
            {
              cmp: cmp,
              data: match$1[1]
            }
          ],
          match[1]
        ];
}

function mergeU(s1, s2, f) {
  var cmp = s1.cmp;
  return {
          cmp: cmp,
          data: Bs_MapDict.mergeU(s1.data, s2.data, f, cmp)
        };
}

function merge(s1, s2, f) {
  return mergeU(s1, s2, Curry.__3(f));
}

function make(id) {
  return {
          cmp: id[/* cmp */0],
          data: Bs_MapDict.empty
        };
}

function isEmpty(map) {
  return Bs_MapDict.isEmpty(map.data);
}

function forEachU(m, f) {
  return Bs_MapDict.forEachU(m.data, f);
}

function forEach(m, f) {
  return Bs_MapDict.forEachU(m.data, Curry.__2(f));
}

function reduceU(m, acc, f) {
  return Bs_MapDict.reduceU(m.data, acc, f);
}

function reduce(m, acc, f) {
  return reduceU(m, acc, Curry.__3(f));
}

function everyU(m, f) {
  return Bs_MapDict.everyU(m.data, f);
}

function every(m, f) {
  return Bs_MapDict.everyU(m.data, Curry.__2(f));
}

function someU(m, f) {
  return Bs_MapDict.someU(m.data, f);
}

function some(m, f) {
  return Bs_MapDict.someU(m.data, Curry.__2(f));
}

function keepU(m, f) {
  return {
          cmp: m.cmp,
          data: Bs_MapDict.keepU(m.data, f)
        };
}

function keep(m, f) {
  return keepU(m, Curry.__2(f));
}

function partitionU(m, p) {
  var cmp = m.cmp;
  var match = Bs_MapDict.partitionU(m.data, p);
  return /* tuple */[
          {
            cmp: cmp,
            data: match[0]
          },
          {
            cmp: cmp,
            data: match[1]
          }
        ];
}

function partition(m, p) {
  return partitionU(m, Curry.__2(p));
}

function mapU(m, f) {
  return {
          cmp: m.cmp,
          data: Bs_MapDict.mapU(m.data, f)
        };
}

function map(m, f) {
  return mapU(m, Curry.__1(f));
}

function mapWithKeyU(m, f) {
  return {
          cmp: m.cmp,
          data: Bs_MapDict.mapWithKeyU(m.data, f)
        };
}

function mapWithKey(m, f) {
  return mapWithKeyU(m, Curry.__2(f));
}

function size(map) {
  return Bs_MapDict.size(map.data);
}

function toList(map) {
  return Bs_MapDict.toList(map.data);
}

function toArray(m) {
  return Bs_MapDict.toArray(m.data);
}

function keysToArray(m) {
  return Bs_MapDict.keysToArray(m.data);
}

function valuesToArray(m) {
  return Bs_MapDict.valuesToArray(m.data);
}

function minKey(m) {
  return Bs_MapDict.minKey(m.data);
}

function minKeyUndefined(m) {
  return Bs_MapDict.minKeyUndefined(m.data);
}

function maxKey(m) {
  return Bs_MapDict.maxKey(m.data);
}

function maxKeyUndefined(m) {
  return Bs_MapDict.maxKeyUndefined(m.data);
}

function minimum(m) {
  return Bs_MapDict.minimum(m.data);
}

function minUndefined(m) {
  return Bs_MapDict.minUndefined(m.data);
}

function maximum(m) {
  return Bs_MapDict.maximum(m.data);
}

function maxUndefined(m) {
  return Bs_MapDict.maxUndefined(m.data);
}

function get(map, x) {
  return Bs_MapDict.get(map.data, x, map.cmp);
}

function getUndefined(map, x) {
  return Bs_MapDict.getUndefined(map.data, x, map.cmp);
}

function getWithDefault(map, x, def) {
  return Bs_MapDict.getWithDefault(map.data, x, def, map.cmp);
}

function getExn(map, x) {
  return Bs_MapDict.getExn(map.data, x, map.cmp);
}

function has(map, x) {
  return Bs_MapDict.has(map.data, x, map.cmp);
}

function checkInvariantInternal(m) {
  return Bs_MapDict.checkInvariantInternal(m.data);
}

function eqU(m1, m2, veq) {
  return Bs_MapDict.eqU(m1.data, m2.data, m1.cmp, veq);
}

function eq(m1, m2, veq) {
  return eqU(m1, m2, Curry.__2(veq));
}

function cmpU(m1, m2, vcmp) {
  return Bs_MapDict.cmpU(m1.data, m2.data, m1.cmp, vcmp);
}

function cmp(m1, m2, vcmp) {
  return cmpU(m1, m2, Curry.__2(vcmp));
}

function getData(prim) {
  return prim.data;
}

function getId(m) {
  var cmp = m.cmp;
  return /* module */[/* cmp */cmp];
}

function packIdData(id, data) {
  return {
          cmp: id[/* cmp */0],
          data: data
        };
}

var Int = 0;

var $$String = 0;

var Dict = 0;

exports.Int = Int;
exports.$$String = $$String;
exports.Dict = Dict;
exports.make = make;
exports.isEmpty = isEmpty;
exports.has = has;
exports.cmpU = cmpU;
exports.cmp = cmp;
exports.eqU = eqU;
exports.eq = eq;
exports.forEachU = forEachU;
exports.forEach = forEach;
exports.reduceU = reduceU;
exports.reduce = reduce;
exports.everyU = everyU;
exports.every = every;
exports.someU = someU;
exports.some = some;
exports.size = size;
exports.toList = toList;
exports.toArray = toArray;
exports.ofArray = ofArray;
exports.keysToArray = keysToArray;
exports.valuesToArray = valuesToArray;
exports.minKey = minKey;
exports.minKeyUndefined = minKeyUndefined;
exports.maxKey = maxKey;
exports.maxKeyUndefined = maxKeyUndefined;
exports.minimum = minimum;
exports.minUndefined = minUndefined;
exports.maximum = maximum;
exports.maxUndefined = maxUndefined;
exports.get = get;
exports.getUndefined = getUndefined;
exports.getWithDefault = getWithDefault;
exports.getExn = getExn;
exports.checkInvariantInternal = checkInvariantInternal;
exports.remove = remove;
exports.removeMany = removeMany;
exports.set = set;
exports.updateU = updateU;
exports.update = update;
exports.mergeU = mergeU;
exports.merge = merge;
exports.mergeMany = mergeMany;
exports.keepU = keepU;
exports.keep = keep;
exports.partitionU = partitionU;
exports.partition = partition;
exports.split = split;
exports.mapU = mapU;
exports.map = map;
exports.mapWithKeyU = mapWithKeyU;
exports.mapWithKey = mapWithKey;
exports.getId = getId;
exports.getData = getData;
exports.packIdData = packIdData;
/* No side effect */
